### PR TITLE
8283092: JMX subclass permission check redundant with strong encapsulation

### PR DESCRIPTION
**Pull request by kevinjwalls**
Removing permission checks which, in the presence of a Security Manager, would check for a RuntimePermission "className.subclass".  This was to prevent subclassing these classes, but is no longer necessary with strong encapsulation from modules.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283092](https://bugs.openjdk.java.net/browse/JDK-8283092): JMX subclass permission check redundant with strong encapsulation

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to a6d05f7367432c0226e79238e909ecc2288e24cc
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7827/head:pull/7827` \
`$ git checkout pull/7827`

Update a local copy of the PR: \
`$ git checkout pull/7827` \
`$ git pull https://git.openjdk.java.net/jdk pull/7827/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7827`

View PR using the GUI difftool: \
`$ git pr show -t 7827`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7827.diff">https://git.openjdk.java.net/jdk/pull/7827.diff</a>

</details>

### CODE REVIEW

File: src/jdk.management.agent/share/classes/jdk/internal/agent/spi/AgentProvider.java [hunk 1]
```diff
@@ -33,27 +33,13 @@
 
     /**
      * Instantiates a new AgentProvider.
```
Comment (dfuch) [inline src/jdk.management.agent/share/classes/jdk/internal/agent/spi/AgentProvider.java:35]: This class should probably be removed altogether. I see that you logged [JDK-8283254](https://bugs.openjdk.java.net/browse/JDK-8283254) so this is fine.

File: src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java [hunk 2]
```diff
@@ -222,13 +218,4 @@ private PlatformMBeanProvider(Void unused) {
...
     public abstract List<PlatformComponent<?>> getPlatformComponentList();
-
-    private static Void checkSubclassPermission() {
-        @SuppressWarnings("removal")
-        SecurityManager sm = System.getSecurityManager();
-        if (sm != null) {
-            sm.checkPermission(new RuntimePermission(PlatformMBeanProvider.class.getName()+".subclass"));
-        }
-        return null;
-    }
```
Comment (dfuch) [inline src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java:233]: I have verified in module-info.java that sun.management.spi is only conditionally exported so I agree that the explicit permission check can now be removed.

**Review by dfuch — APPROVED** [review]

File: src/jdk.management.agent/share/classes/jdk/internal/agent/spi/AgentProvider.java (hunk 1 above)
Comment (kevinjwalls) [reply src/jdk.management.agent/share/classes/jdk/internal/agent/spi/AgentProvider.java:35]: Thanks Daniel - yes will get rid of this class soon, I'll keep these two actions in separate changes. 8-)

File: src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java (hunk 2 above)
Comment (kevinjwalls) [reply src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java:233]: thanks

**Comment by seanjmullan** [comment]
I think you also need to update this test such that it checks that the module access restrictions prevent subclassing instead of the permission check: test/jdk/sun/management/PlatformMBeanProviderConstructorCheck.java

Also, it looks like this permission was never publicly documented and was only used internally, so it probably doesn't need a CSR, but I don't know for sure.

**Comment by kevinjwalls** [comment]
> test/jdk/sun/management/PlatformMBeanProviderConstructorCheck.java

Thank for noticing that Sean - had run various tests but missed this.  I have an update, will add it here soon.

**Comment by kevinjwalls** [comment]
Added test update.  Checks that with permissive module options we can extend sun.management.spi.PlatformMBeanProvider, and that without them we cannot.

Output of the new test is like this:

----------System.out:(11/1279)----------
---PlatformMBeanProviderConstructorCheck:
---PlatformMBeanProviderConstructorCheck: invoke MyProvider with expectedFail = false
---PlatformMBeanProviderConstructorCheck PASSED (1) (expectedFail = false)
---PlatformMBeanProviderConstructorCheck: re-invoke without --add-modules or --add-exports
Command line: [/tank/kwalls/repos/personal/jdk/open/test/../../build/linux-x86_64-server-release/images/jdk/bin/java -cp /tank/kwalls/repos/personal/jdk/open/test/JTwork/classes/sun/management/PlatformMBeanProviderConstructorCheck.d:/tank/kwalls/repos/personal/jdk/open/test/jdk/sun/management:/tank/kwalls/repos/personal/jdk/open/test/JTwork/classes/test/lib:/tank/kwalls/repos/personal/jdk/open/test/lib:/opt/jtreg6.1/lib/javatest.jar:/opt/jtreg6.1/lib/jtreg.jar PlatformMBeanProviderConstructorCheck --nomoduleargs ]
[2022-03-18T17:19:34.798024163Z] Gathering output for process 10627
[2022-03-18T17:19:34.902532753Z] Waiting for completion for process 10627
[2022-03-18T17:19:34.902807078Z] Waiting for completion finished for process 10627
Output and diagnostic info for process 10627 was saved into 'pid-10627-output.log'
[2022-03-18T17:19:34.908368606Z] Waiting for completion for process 10627
[2022-03-18T17:19:34.908503964Z] Waiting for completion finished for process 10627
----------System.err:(9/647)----------
 stdout: [---PlatformMBeanProviderConstructorCheck:
---PlatformMBeanProviderConstructorCheck: invoke MyProvider with expectedFail = true
---PlatformMBeanProviderConstructorCheck got exception: java.lang.IllegalAccessError: superclass access check failed: class PlatformMBeanProviderConstructorCheck$MyProvider (in unnamed module @0x4795bbf5) cannot access class sun.management.spi.PlatformMBeanProvider (in module java.management) because module java.management does not export sun.management.spi to unnamed module @0x4795bbf5
---PlatformMBeanProviderConstructorCheck PASSED (2) (expectedFail = true)
];
 stderr: []
 exitValue = 0

STATUS:Passed.

File: src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java [hunk 3]
```diff
@@ -203,12 +203,8 @@ public default boolean shouldRegister() {
...
      * Instantiates a new PlatformMBeanProvider.
-     *
-     * @throws SecurityException if the subclass (and calling code) does not
-     *    have {@code RuntimePermission("sun.management.spi.PlatformMBeanProvider.subclass")}
      */
     protected PlatformMBeanProvider () {
-        this(checkSubclassPermission());
     }
 
     private PlatformMBeanProvider(Void unused) {
```
Comment (mlchung) [inline src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java:210]: This `PlatformMBeanProvider(Void)` constructor is no longer needed.   Same for `AgentProvider(Void)` constructor.

Comment (kevinjwalls) [reply src/java.management/share/classes/sun/management/spi/PlatformMBeanProvider.java]: Thanks, yes they can go...  Rebuilt and tested ok after the new commit.

**Review by mlchung — APPROVED** [review]

**Comment by kevinjwalls** [comment]
Thanks all for the comments and reviews!

**Pull request closed without merging**